ray_on_golem/server/cluster/cluster.py

```diff
@@ -0,0 +1,228 @@
+
+from ray_on_golem.server.cluster.nodes import ClusterNode, HeadClusterNode, WorkerClusterNode
+from ray_on_golem.server.mixins import WarningMessagesMixin
+from ray_on_golem.server.models import (
+    NodeConfigData,
+    NodeId,
+    NodeState,
+    ProviderParametersData,
+    Tags,
+)
+from ray_on_golem.server.services.golem.golem import DeviceListAllocationPaymentManager
+from ray_on_golem.server.services.golem.helpers.demand_config import DemandConfigHelper
+from ray_on_golem.server.services.golem.manager_stack import ManagerStack
+from ray_on_golem.server.services.new_golem import GolemService
+
+IsHeadNode = bool
+NodeType = str
+StackKey = Tuple[NodeType, IsHeadNode]
+
+logger = logging.getLogger(__name__)
+
+
+class Cluster(WarningMessagesMixin):
+    def __init__(
+        self,
+        golem_service: GolemService,
+        name: str,
+        provider_parameters: ProviderParametersData,
+        webserver_port: int,
+        demand_config_helper: DemandConfigHelper,
+    ) -> None:
+        super().__init__()
+
+        self._golem_service = golem_service
+        self._name = name
+        self._provider_parameters = provider_parameters
+        self._webserver_port = webserver_port
+        self._demand_config_helper = demand_config_helper
+
+        self._manager_stacks: Dict[StackKey, ManagerStack] = {}
+        self._manager_stacks_locks: DefaultDict[StackKey, asyncio.Semaphore] = defaultdict(
+            asyncio.Semaphore
+        )
+        self._nodes: Dict[NodeId, ClusterNode] = {}
+        self._nodes_id_counter = 0
+        self._payment_manager: PaymentManager = DeviceListAllocationPaymentManager(
+            self._golem_service.golem,
+            budget=self._provider_parameters.total_budget,
+            network=self._provider_parameters.payment_network,
+            driver=self._provider_parameters.payment_driver,
+        )
+
+        self._state: NodeState = NodeState.terminated
+
+    @property
+    def nodes(self) -> Mapping[str, ClusterNode]:
+        return self._nodes
+
+    def get_warning_messages(self) -> List[str]:
+        warnings = super().get_warning_messages()
+
+        for node in self._nodes.values():
+            warnings.extend(node.get_warning_messages())
+
+        return warnings
+
+    async def start(self) -> None:
+        if self._state in (NodeState.pending, NodeState.running):
+            logger.info("Not starting `%s` cluster as it's already running or starting", self._name)
+            return
+
+        logger.info("Starting `%s` cluster...", self._name)
+
+        self._state = NodeState.pending
+
+        await self._payment_manager.start()
+
+        self._state = NodeState.running
+
+        logger.info("Starting `%s` cluster done", self._name)
+
+    async def stop(self) -> None:
+        if self._state in (NodeState.terminating, NodeState.terminated):
+            logger.info("Not stopping `%s` cluster as it's already stopped or stopping", self._name)
```

Comment:
I'm afraid that I need to disagree. Here, the log is related to the actual condition. Your proposal can be misleading in case the cluster was never started.

Reply:
well, the message above _also_ implies that the cluster has been started at some point...